internal/aeintegrate/aeintegrate.go

```diff
@@ -0,0 +1,350 @@
+		}
+	}
+
+	out, err := yaml.Marshal(c)
+	if err != nil {
+		return "", err
+	}
+
+	if err := ioutil.WriteFile(filepath.Join(p.Dir, tmp), out, 0755); err != nil {
+		return "", err
+	}
+
+	p.tempAppYaml = tmp
+	return p.tempAppYaml, nil
+}
+
+func (p *App) deployCmd() (*exec.Cmd, error) {
+	gcloudBin := os.Getenv("GCLOUD_BIN")
+	if gcloudBin == "" {
+		gcloudBin = "gcloud"
+	}
+	aedeploy := os.Getenv("AEDEPLOY_BIN")
+	if aedeploy == "" {
+		aedeploy = "aedeploy"
+	}
+
+	appyaml, err := p.envappyaml()
+	if err != nil {
+		return nil, err
+	}
+
+	// NOTE: if the "preview" and/or "app" components are not available, and this is
+	// run in parallel, gcloud will attempt to install those components multiple
+	// times and will eventually fail on IO.
+	args := []string{gcloudBin,
+		"--quiet",
+		"preview", "app", "deploy", appyaml,
+		"--project", p.ProjectID,
+		"--version", p.version(),
+		"--no-promote",
+	}
+	cmd := exec.Command(aedeploy, args...)
+	cmd.Dir = p.Dir
+	return cmd, nil
+}
+
+// Module returns the Module ID, which is read from the app.yaml file.
+func (p *App) Module() (string, error) {
+	if p.module != "" {
+		return p.module, nil
+	}
+
+	b, err := ioutil.ReadFile(filepath.Join(p.Dir, p.appyaml()))
+	if err != nil {
+		return "", err
+	}
+
+	s := &struct {
+		Module string `yaml:"module"`
+	}{}
+
+	if err := yaml.Unmarshal(b, s); err != nil {
+		return "", err
+	}
+
+	if s.Module == "" {
+		s.Module = "default"
+	}
+
+	p.module = s.Module
+	return p.module, err
+}
+
+// setupAdmin populates p.adminService and checks that the user is authenticated and project ID is valid.
+func (p *App) setupAdmin() error {
+	c, err := google.DefaultClient(context.Background(), appengine.CloudPlatformScope)
+	if err != nil {
+		return err
+	}
+
+	if p.adminService, err = appengine.New(c); err != nil {
+		return err
+	}
+
+	if err := p.validate(); err != nil {
+		return err
+	}
+
+	// Check that the user is authenticated, etc.
+	_, err = p.adminService.Apps.Get(p.ProjectID).Do()
+	return err
+}
+
+// Cleanup deletes the created version from App Engine.
+func (p *App) Cleanup() error {
+	// NOTE: don't check whether p.deployed is set.
+	// We may want to attempt to clean up if deployment failed.
+	// However, we require adminService to be set up, which happens during Deploy().
+	if p.adminService == nil {
+		return errors.New("Cleanup called before Deploy")
+	}
+
+	if err := p.validate(); err != nil {
+		return err
+	}
+
+	if p.tempAppYaml != "" {
+		if err := os.Remove(filepath.Join(p.Dir, p.tempAppYaml)); err != nil {
+			// Continue trying to clean up, even if the temp yaml file didn't get removed.
+			log.Print(err)
+		}
+	}
+
+	log.Printf("(%s) Cleaning up.", p.Name)
+
+	var err error
+	for try := 0; try < 10; try++ {
+		_, err = p.adminService.Apps.Modules.Versions.Delete(p.ProjectID, p.module, p.version()).Do()
+		if err == nil {
+			log.Printf("(%s) Succesfully cleaned up.", p.Name)
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	return err
```

Comment:
fmt.Errorf("could not delete app module version %v/%v: %v", p.module, p.version(), err)